Bluetooth: Mesh: Add support prioritized message relay

LingaoM opened the pull request:
For Mesh Message, should only process by model layer when dst is unicast address of this nodes or model subscrip on.

The prioritized relay messages will be send before other lower
priority relay message to ensure that higher priority messages
can be sent in time. When considering the message latency, also
consider the values of BT_MESH_RELAY_RETRANSMIT_COUNT and
BT_MESH_RELAY_RETRANSMIT_INTERVAL.

A higher number of BT_MESH_RELAY_ADV_SETS allows the increase in
the number of buffers while maintaining the latency.

Note: To enable this option, the application layer needs to register
a callback function to determine the priority of each relay message.

If it is not registered, the default priority is the same as the lowest.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

PavelVPV commented on subsys/bluetooth/mesh/transport.c:
```diff
@@ -982,6 +982,10 @@ static int trans_unseg(struct net_buf_simple *buf, struct bt_mesh_net_rx *rx,
 		return -EINVAL;
 	}
 
+	if (!rx->local_match) {
```
Could this check be removed?
https://github.com/zephyrproject-rtos/zephyr/blob/cd3c3cf6ebc136e8672220382a0c59622b51531b/subsys/bluetooth/mesh/transport.c#L756-L760

LingaoM replied:
Thanks

LingaoM replied:
Move to this more suitable.

PavelVPV reviewed (commented):
This change broke 3 bsim tests

LingaoM committed:
Bluetooth: Mesh: Add support prioritized message relay

In our use scenario, the gateway needs to control a large
number of light bulbs in real time, which can exceed 100,
and a large number of light bulbs will return the status to
the user every time the status changes, which will cause a
temporary mesh network storm in the whole network.

We optimized the time of status return, which does not happen
immediately, but will return the status for a random number of
seconds, but during the status return, Each status message may
be relayed by all nodes in the entire network.

If the gateway sends a new control message at this time, the
message obviously needs to be queued, resulting in poor consistency
of the entire network control.

The prioritized relay messages will be send before other lower
priority relay message to ensure that higher priority messages
can be sent in time. When considering the message latency, also
consider the values of BT_MESH_RELAY_RETRANSMIT_COUNT and
BT_MESH_RELAY_RETRANSMIT_INTERVAL.

A higher number of BT_MESH_RELAY_ADV_SETS allows the increase in
the number of buffers while maintaining the latency.

Note: To enable this option, the application layer needs to register
a callback function to determine the priority of each relay message.

If it is not registered, the default priority is the same as the lowest.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
> This change broke 3 bsim tests

All tests passed, Please take a look.

PavelVPV reviewed (changes requested):
@LingaoM with which parameters of the stack (network transmit and retransmit states, number of relay adv sets, adv buf count, etc.) did you measure the relay node performance and did you measure the latency on the relay node?

LingaoM commented:
Sure，The delay(maximum delay almost 600ms) of the relay depends on the number of relay buffers(currently 12)and the speed of each transmission ( 3 * 50ms ) and the number of parallel relays( 2 relay adv sets).

alxelax commented on subsys/bluetooth/mesh/adv.h:
```diff
@@ -47,9 +48,10 @@ struct bt_mesh_adv {
+struct net_buf *bt_mesh_adv_main_create(enum bt_mesh_adv_type type,
+					uint8_t xmit, k_timeout_t timeout);
+
+struct net_buf *bt_mesh_adv_relay_create(bool hp, uint8_t xmit);
```
It will be nice to have a linked list sorted according to priority and more granularity for priorities. The problem with inefficient usage of two FIFOs will go away. Priority will be always actual. All available memory will be used in an efficient manner. Customers can customize priorities for different frames.

LingaoM replied:
it's a good idea

LingaoM replied:
My concern is that the need to dynamically adjust the list each time may lead to cpu overhead

LingaoM replied:
Currently, two pool method, always to ensure high priority messages have buffer to relay, instead of all being consumed by low-priority messages.

LingaoM replied:
Yes, maybe we can remove from list, the lowest priority messages will be free for higher priority messages.

LingaoM replied:
This frequent traversal of the list needs to be thread-safe.

LingaoM replied:
Although sysworkq cannot be interrupted, application tasks may be preempted, so race conditions may occur

alxelax replied:
Could `SYS_SLIST_FOR_EACH_CONTAINER_SAFE` help to avoid race conditions?

LingaoM replied:
No, I used mutex for thread protection.

LingaoM replied:
I also don't delete nodes in the list traversal. I'm worried about different tasks accessing the queue, so it is possible to lock through mutex.

LingaoM replied:
![image](https://user-images.githubusercontent.com/26378606/201298498-6e483bd4-6d52-4204-bcf0-708f4dbd3898.png)

All tests great, maybe we need add new testcase for this priority relay.

But it's better to do this after you two have finished the review. I don't want to do it in rebase.

LingaoM replied:
One important problem found.

net_buf_unref will destroy the nodes behind the queue.

alxelax reviewed (changes requested):
Agree with @PavelVPV. Seems it requires more work to handle correctly corner cases.
Probably dual priority is not sufficient.

LingaoM commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -380,6 +380,17 @@ endchoice
 
+config BT_MESH_ADV_MAIN_RELAY_ENABLED
+	bool "Advertiser main support relay messages"
+	default y
```
That true, thanks

LingaoM commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -93,24 +103,42 @@ static struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *buf_poo
+		buf = bt_mesh_adv_create_from_pool(&relay_buf_hp_pool, adv_relay_hp_pool,
+						   BT_MESH_ADV_DATA, BT_MESH_RELAY_ADV,
+						   true, xmit, K_NO_WAIT);
+		if (buf) {
+			return buf;
+		}
```
I only agree with some opinions. First of all, if the high-priority messages are too dense, resulting in insufficient buffers in the high-priority pool, then the message will be backed up as a normal relay message. This is the current practice, and what you mean is that even from  If the message is obtained from the ordinary relay pool, it can still be placed in the header of the relay_queue, so that the message can be relayed as quickly as possible. This is a good idea.

But I guess there is no worst case here, because for broadcast set it always takes priority to send messages from high priority.

LingaoM replied:
> The worst case is when the prioritized pool never gets empty and the messages from the non-prioritized pool are never sent.

I think this has exceeded the original intention of high priority. My expectation is to customize a limited buffer to send relays quickly, but if there are dense messages that require high-priority relays, then it is actually back to our problem.  early.

PavelVPV commented:
> Sure，Before this pr, the delay(maximum delay almost 600ms) of the relay depends on the number of relay buffers(currently 12)and the speed of each transmission ( 3 * 50ms ) and the number of parallel relays( 2 relay adv sets and one main adv).
> 
> After this pr, the high priority messages will relay as far as possible, also if advertiser main only relay high priority message, this this delay will be less than 20ms, based on other relay adv sets all busy to relay general messages.

Can this be solved using currently available configuration?

For example, to reduce latency, `CONFIG_BT_MESH_RELAY_BUF_COUNT` was introduced, which value can be calculated as following:
```
CONFIG_BT_MESH_RELAY_BUF_COUNT = (latency [ms] / adv duration [ms]) * CONFIG_BT_MESH_RELAY_ADV_SETS
```
where:
- `latency` - the desired latency
- `adv duration` (the worst case) defined here: https://github.com/zephyrproject-rtos/zephyr/blob/f98a240537f34da826f4dc6eb95f63b4e8bb845b/subsys/bluetooth/mesh/adv_ext.c#L178 

But adv duration is usually much less than the worst case, especially if you set Network Transmit Count and Relay Retransmit Count states to 0. So, for example, in nRF Connect SDK to achieve 50ms latency on a relay node we use ~1.5 messages per adv set with Relay Retransmit Count state set to 0. 

Then, increasing `CONFIG_BT_MESH_RELAY_ADV_SETS` you can find a balance between reception and transmission of messages. So, for example, in nRF Connect SDK we use 6 adv sets to achieve ~120-130 messages per second on a relay node without packet drops due to small `CONFIG_BT_MESH_RELAY_BUF_COUNT`.

LingaoM commented:
> Can this be solved using currently available configuration?

Unfortunately, if the priority is not adopted, the relay effect of the control message sent by the gateway is not good because there is no available relay buffer. This is the original intention of this PR. After priority is adopted, high-priority buffers can always preempt low-priority relay messages, which completely guarantees the relay experience.

PavelVPV commented:
> > Can this be solved using currently available configuration?
> 
> Unfortunately, if the priority is not adopted, the relay effect of the control message sent by the gateway is not good because there is no available relay buffer. This is the original intention of this PR. After priority is adopted, high-priority buffers can always preempt low-priority relay messages, which completely guarantees the relay experience.

Could you please try to check with the following configuration?
```
CONFIG_BT_MESH_RELAY_ADV_SETS=6 # though 4 or 5 could be enough for ZLL
CONFIG_BT_MESH_RELAY_BUF_COUNT=10
CONFIG_BT_MESH_RELAY_RETRANSMIT_COUNT=0
CONFIG_BT_MESH_NETWORK_TRANSMIT_COUNT=0
```

LingaoM commented:
> Could you please try to check with the following configuration?
> 
> ```
> CONFIG_BT_MESH_RELAY_ADV_SETS=6 # though 4 or 5 could be enough for ZLL
> CONFIG_BT_MESH_RELAY_BUF_COUNT=10
> CONFIG_BT_MESH_RELAY_RETRANSMIT_COUNT=0
> CONFIG_BT_MESH_NETWORK_TRANSMIT_COUNT=0
> ```

@PavelVPV In our application scenario, the parameters of relay are fixed. For some historical reasons, we cannot change these parameters rashly.

PavelVPV commented:
> > Could you please try to check with the following configuration?
> > ```
> > CONFIG_BT_MESH_RELAY_ADV_SETS=6 # though 4 or 5 could be enough for ZLL
> > CONFIG_BT_MESH_RELAY_BUF_COUNT=10
> > CONFIG_BT_MESH_RELAY_RETRANSMIT_COUNT=0
> > CONFIG_BT_MESH_NETWORK_TRANSMIT_COUNT=0
> > ```
> 
> @PavelVPV In our application scenario, the parameters of relay are fixed. For some historical reasons, we cannot change these parameters rashly.

But can you increase number of relay adv sets considering time needed for one advertiser to relay a message?

How do you ensure in your use case that the right messages get higher prioritization? Unless I'm missing something, network layer won't give enough information for proper prioritization of access messages.

LingaoM commented:
> How do you ensure in your use case that the right messages get higher prioritization? Unless I'm missing something, network layer won't give enough information for proper prioritization of access messages.

Currently I'm according to dst field.

omkar3141 commented:
Hi Lingao,
I'm afraid that this is introducing a change to the mesh stack that can potentially have unintended consequences for regular mesh operation. If the node implementing this feature is used in an interoperable-mixed-mesh-eco-system (which has nodes/relays from many different vendors), then these nodes will have different behavior. On account of the prioritization of some messages, the latency of some other messages (say, lighting relayed on-off messages) will suffer negatively, or worse, messages may experience a higher drop rate and reduced reliability. For example, if you have more than 5 transport control messages arriving quickly within a couple of milliseconds and the node has a total of 5 ADV sets, all of these will be prioritized and pushed into the relay queue. If other lighting-related messages arrive at a close-by time, those messages will get dropped as the queue may have gotten full due to high-priority messages. Additionally, if you get other high-priority messages at the same time, it will also cause them to be dropped or delayed invariably until previous messages are out (depending on the choice of implementation). So, this feature can only work if high-priority messages have fairly low traffic (say once in a few seconds), in other cases this feature has the potential of affecting the whole network. 

Looking at the PR description it seems like this is trying to solve the implementation-specific issue where wrong assumptions are used in implementing the lighting control system. A properly implemented mesh network would not need "Status" messages from each light to be reaching the gateway. It is possible to achieve 100% success in turning lights on-off with the use of unacknowledged messages. Putting the requirement of sending statuses back is making the network worse. No amount of prioritization will solve this problem. As your gateway starts inserting more higher-priority messages, the network will get flooded more with them and lead to worse latency or many packet drops for other messages as well.

> If the gateway sends a new control message at this time, the message obviously needs to be queued, resulting in poor consistency of the entire network control.

I kind of disagree with this part as well. When messages are injected into the network they will be picked up by neighboring relays. Careful choice of network retransmit count and relay retransmit count and intervals during the network planning and commissioning phase can solve this problem easily. Because of the flooding nature of the Bluetooth mesh, at some point, the network will always have collisions and packet drops, and delays. The Bluetooth mesh works best when there is the least amount of traffic in the network. The network planning and conceptualization should be directed toward minimizing traffic. You won't need status messages from all lights back to Gateway if Gateway's location is central and the number of relays is limited in the network.

> Note: To enable this option, the application layer needs to register a callback function to determine the priority of each relay message.

Having a callback in the relay mechanism to give the application layer some say over packet priority or packet rejection is a useful mechanism but inserting this in the stack without any default way to keep it disabled will make the stack will behave very differently compared to other mesh nodes that do not have this. Ideally, this feature should be disabled by default and should only be enabled explicitly via some kind of vendor model. 

The changes in the adv.c relay logic seem very application specific. I suggest you can maintain them as a patch on your side instead of integrating in the main stack. What if someone wants to use relay callback to decide something else other than priority (for example, skipping a message, or generating a new one)? There is no such option possible. If new need arises, will we have to change this API at that time? would it break backwards compatibility then?

LingaoM commented:
Hi @omkar3141 

 > I'm afraid that this is introducing a change to the mesh stack that can potentially have unintended consequences for regular mesh operation.

Regarding the priority relay, in fact, the spec has no clear restrictions/definitions, and I don't think it breaks the standard consistency. Of course, we need sufficient testing to ensure a reliable experience.

> if you have more than 5 transport control messages arriving quickly within a couple of milliseconds and the node has a total of 5 ADV sets, all of these will be prioritized and pushed into the relay queue. If other lighting-related messages arrive at a close-by time, those messages will get dropped as the queue may have gotten full due to high-priority messages. 

According to my discussion yesterday, the dual-pool mode has been removed, and messages with higher priority can directly replace messages with lower priority if relay buffer insufficient, so there is no packet loss problem you mentioned.

> You won't need status messages from all lights back to Gateway if Gateway's location is central and the number of relays is limited in the network.

I fully agree that our current test is very extreme. This is still carried out in the case of 0 to 10 seconds of status return. Of course, this is also my thinking. For the large number of useless relay messages in the network, it is the same as the possession of control messages.  The scheduling resources are actually a waste, which is also my original intention of doing this PR.

> Having a callback in the relay mechanism to give the application layer some say over packet priority or packet rejection is a useful mechanism but inserting this in the stack without any default way to keep it disabled will make the stack will behave very differently compared to other mesh nodes that do not have this. Ideally, this feature should be disabled by default and should only be enabled explicitly via some kind of vendor model.

That's actually the case, I hope this only affects vendor messages and is off by default.  This needs a special statement, thank you for the reminder.

> The changes in the adv.c relay logic seem very application specific. I suggest you can maintain them as a patch on your side instead of integrating in the main stack. What if someone wants to use relay callback to decide something else other than priority (for example, skipping a message, or generating a new one)? There is no such option possible. If new need arises, will we have to change this API at that time? would it break backwards compatibility then?

My original intention is to keep the main branch as consistent as possible, so as to reduce our maintenance burden. Of course, all this has to refer to your opinions. I hope to do this together, instead of giving up because of some concerns, right?  Finally, regarding the backward compatibility of the API, perhaps the skip relay message you mentioned does not conform to the specification of the Mesh Profile. Of course, there may be new requirements, but if there is a rationality, we can discuss it.


The above is my opinion, thank you very much!

LingaoM committed:
tests: bluetooth: mesh: Add tests for priority relay

Add init support for priority relay.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
Hi, @PavelVPV 

It's also a compromise, and to me, your requests for revisions make sense, and I'm happy to accept them, so you don't have to be sorry.

LingaoM commented:
Hi @omkar3141 

In addition, in general, messages in the network should have QoS flags, similar to TCP/IP or Bluetooth L2CAP Channels, to manage the priority of each message, even if it is not defined in the current Mesh Profile, but  By no means is this way wrong, so I hope you'll be able to discuss this with the Bluetooth SIG Mesh Group, I believe it's a nice performance boost.

LingaoM commented:
I thought about it carefully, I don't think this PR may break the standard interaction consistency as @omkar3141 mentioned.

The reason is: from the perspective of wireless transmission, priority relay is actually equivalent to infinite multiple advertising sets at the same time relay, obviously for multiple advertising sets, there are no relevant constraints on the relay protocol, because this does not break anything, but improves the interactive experience.  

In addition, regarding what @PavelVPV  mentioned, this PR is too specific for specific application scenarios. This PR does not help the current protocol stack. 

I agree with this point of view from a standard perspective, but I think this PR is a the new relay implementation method, it tries to tap its performance as much as possible within the scope allowed by the protocol.

Also the current priority method has exceeded the optimization of the specific scenario i mentioned at the beginning, and now this pr is more general. It can solve things that FIFO relays can't solve in specific scenarios, just like we use multi-advertising sets relays to solve things that legacy advertising can't solve, right?

PavelVPV commented:
Looking at the current changes again I still don't see a reason of adding this to the main branch. With the parameters I mentioned earlier (https://github.com/zephyrproject-rtos/zephyr/pull/52072#issuecomment-1310871526) in our tests we can achieve pretty good relay node performance with ZLL with almost zero packet loss (due to small relay buffer). So I do believe at this moment that with proper configuration of the stack the prioritization is not needed. As I mentioned earlier, if there is a way to let a user use own or redefine the relay buffer in a way that this change won't harm the current code, this is something that could be accepted.

LingaoM commented:
This may not be very useful for Nordic's Controller, but it is also a better solution for a limited number of multicast sets. Most of the Vendors we have used only have four multicast sets, one of the broadcast sets can only be used as Proxy Advertising.

mia-ko commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -380,6 +380,17 @@ endchoice
+	default y if !BT_MESH_FRIEND && !BT_MESH_LOW_POWER
+	depends on BT_MESH_RELAY
+	help
+	  Enable advertiser main to support relay messages, this maximizes the
+	  utilization efficiency of advertising sets, which is helpful for the
+	  sending of dense relays, but for Friend and LPN, it delays the sending
+	  of local messages a little, this may cause the time slot deviation between
+	  friend and lpn, and the power consumption of lpn will increase.
```
```suggestion
	  Enable advertiser main to support relay messages, this maximizes the
	  utilization efficiency of advertising sets, which is helpful for the
	  sending of dense relays, but for Friend and LPN, it delays the sending
	  of local messages a little, this may cause the time slot deviation between
	  friend and lpn, and the power consumption of lpn will increase.
```

PavelVPV commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -93,35 +95,126 @@ static struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *buf_poo
+	 * For the same low priority message, the one that takes the longest buffer time will
+	 * be released first.
+	 */
+	k_mutex_lock(&bt_mesh_relay_mutex, K_FOREVER);
```
I don't like to have mutex with sleep at the bottom layer. The mesh code is not thread-safe. If reschedule happens and we enter mesh code from another thread (this is currently possible because ble host calls mesh from a separate thread: https://github.com/zephyrproject-rtos/zephyr/blob/87a751c266da02b53193852e336c3227bf7a3672/subsys/bluetooth/host/Kconfig#L86-L89), this can mess up states.

LingaoM replied:
I also thought about this situation, but ensuring thread safety is a problem. In the design of zephyr bluetooth, it is a conventional limitation to use Co-Operative.

``` C
	k_thread_create(&tx_thread_data, tx_thread_stack,
			K_KERNEL_STACK_SIZEOF(tx_thread_stack),
			hci_tx_thread, NULL, NULL, NULL,
			K_PRIO_COOP(CONFIG_BT_HCI_TX_PRIO),
			0, K_NO_WAIT);
	k_thread_name_set(&tx_thread_data, "BT TX");

#if defined(CONFIG_BT_RECV_WORKQ_BT)
	/* RX thread */
	k_work_queue_init(&bt_workq);
	k_work_queue_start(&bt_workq, rx_thread_stack,
			   CONFIG_BT_RX_STACK_SIZE,
			   K_PRIO_COOP(CONFIG_BT_RX_PRIO), NULL);
	k_thread_name_set(&bt_workq.thread, "BT RX");
#endif
```

LingaoM replied:
So there is no need for mutex before.

PavelVPV commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -93,35 +95,126 @@ static struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *buf_poo
+	 */
+	k_mutex_lock(&bt_mesh_relay_mutex, K_FOREVER);
+
+	SYS_SFLIST_FOR_EACH_NODE(&bt_mesh_relay_queue.data_q, curr) {
```
I guess it is not good to access internal entries of `k_queue`. Perhaps there could be something like `K_QUEUE_FOR_EACH`?

LingaoM replied:
Your suggestion is meaningful.

PavelVPV replied:
I think this needs to be addressed.

PavelVPV commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -157,25 +251,52 @@ struct net_buf *bt_mesh_adv_buf_get(k_timeout_t timeout)
+	/* The messages with the highest priority are always placed at the head, and
+	 * the messages with the same priority are arranged in chronological order.
+	 */
+	k_queue_insert(&bt_mesh_relay_queue, prev, net_buf_ref(buf));
```
I think this should be part of `net_buf` API to guarantee that nothing brakes if for some reason `net_buf` implementation changes. Same for `k_queue_get`.

LingaoM replied:
I'm not sure if this breaks net_ Buf design, net_ Buf inherits FIFO, which is obviously a first in, first out design.

PavelVPV replied:
To be sure that it won't break, I think this needs to be part of `net_buf` API.

LingaoM replied:
I thought about it again, your purpose is to maintain the insert operation in the net_buf API, but what needs to be considered is that the current net_buf is based on FIFO, I don't want to destroy this structure, my final solution is Abandon net_buf and use memslab, so that node nodes can be managed by us, and part of the ram resource overhead can also be solved.

LingaoM replied:
Well, done now.

PavelVPV commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -157,25 +251,52 @@ struct net_buf *bt_mesh_adv_buf_get(k_timeout_t timeout)
 		return bt_mesh_adv_buf_get(timeout);
 	} else if (tag & BT_MESH_RELAY_ADV) {
-		return bt_mesh_adv_buf_relay_get(timeout);
+		return bt_mesh_adv_buf_relay_get();
```
Why timeout argument omitted here?

PavelVPV commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -40,8 +40,9 @@ const uint8_t bt_mesh_adv_type[BT_MESH_ADV_TYPES] = {
 
-static K_FIFO_DEFINE(bt_mesh_adv_queue);
-static K_FIFO_DEFINE(bt_mesh_relay_queue);
+static K_QUEUE_DEFINE(bt_mesh_adv_queue);
+static K_QUEUE_DEFINE(bt_mesh_relay_queue);
```
Though we are planning to add a separate adv set for Friend for better friendship performance, it is actually good idea to use `K_QUEUE` as we could use this to put Friend Update messages in front of the local queue so they are not delayed by other local messages when 1 adv set is used for local messages and friendship.

PavelVPV commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -94,24 +95,84 @@ static struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *buf_poo
 
-	return bt_mesh_adv_create_from_pool(&adv_buf_pool, adv_local_pool, type,
-					    tag, xmit, timeout);
+	sys_sflist_remove(&bt_mesh_relay_queue.data_q, lowest_prev, lowest);
```
`k_queue_remove`?

LingaoM replied:
I don't like this API, it will traverse again, obviously, we don't need additional traverse

LingaoM replied:
I did as you suggested, just because I don't want to add a kernel API,

LingaoM committed:
kernel: add foreach queue macro

Add support foreach queue nodes.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Replace net_buf to memslab

Since net_buf api not support insert operation,
also for save some ram footprint, use memslab may be
suitable.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -41,88 +41,158 @@ const uint8_t bt_mesh_adv_type[BT_MESH_ADV_TYPES] = {
 
-	buf = net_buf_alloc(buf_pool, timeout);
-	if (!buf) {
+	err = k_mem_slab_alloc(buf_pool, (void **)&buf, timeout);
```
Before:
![2MqduCiQoG](https://user-images.githubusercontent.com/26378606/205832501-cafb7155-c4a9-4d86-9919-09cd1e2cb169.jpg)
After:
![image](https://user-images.githubusercontent.com/26378606/205832613-88bb4a3d-d26d-46bc-8799-17c3be1d8e8a.png)

PavelVPV reviewed (changes requested):
Ok, now I think I like this more, though there are few major issues with preempting buffers. It would be also ideal if the commit with switching to `k_mem_slab` would be added before the prioritization change.

LingaoM commented:
>  It would be also ideal if the commit with switching to k_mem_slab would be added before the prioritization change.

I am thinking about the suitability of this matter, but it is also a thorny rebase for me. I think it makes no difference to the final result. Therefore, let's keep the status quo, but we can still do better next time, depending on more full and detailed preliminary consideration.

LingaoM commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -41,88 +41,157 @@ const uint8_t bt_mesh_adv_type[BT_MESH_ADV_TYPES] = {
+	buf = CONTAINER_OF(lowest, struct bt_mesh_buf, node);
+
+	bt_mesh_adv_send_start(0, -ECANCELED, &buf->adv);
+	bt_mesh_buf_unref(buf);
```
For first question:

PavelVPV replied:
Right, I've missed that, thanks! The first point is clear now.

LingaoM replied:
For second question:
When sending a message in adv_ext.c, will dequeue from queue, so this adv buf will not be trace here.
And
For proxy_msg.c, good point, i'm missing this condition

LingaoM replied:
I thought about it carefully, the second problem you mentioned about proxy relay, the final reason may be that we confuse the concept of relay, in the profile of Mesh1.0, relay in many scenarios is defined, I think this is not Should not be addressed in this PR.

LingaoM replied:
### Mesh Profile 1.0.1 3.4.6.3 Receiving a Network PDU

If the message delivered from the advertising bearer is processed by the lower transport layer, the Relay
feature is supported and enabled, the TTL field has a value of 2 or greater, and the destination is not a
unicast address of an element on this node, then the TTL field value shall be decremented by 1, the
Network PDU shall be **tagged as relay**, and the Network PDU shall be retransmitted to all network
interfaces connected to the advertising bearer. It is recommended that a small random delay is introduced
between receiving a Network PDU and relaying a Network PDU to avoid collisions between multiple
relays that have received the Network PDU at the same time.

If the message delivered from the GATT bearer is processed by the lower transport layer, and the Proxy
feature is supported and enabled, and the TTL field has a value of 2 or greater, and the destination is not
a unicast address of an element on this node, then the TTL field value shall be decremented by 1, and
the Network PDU shall be retransmitted to all network interfaces.

If the message delivered from the advertising bearer is processed by the lower transport layer, and the
Proxy feature is supported and enabled, and the TTL field has a value of 2 or greater, and the destination
is not a unicast address of an element on this node, then the TTL field shall be decremented by 1 and the
Network PDU shall be retransmitted to all network interfaces connected to the GATT bearer.

LingaoM replied:
Also done now.

LingaoM replied:
@PavelVPV  All comments have been completed. Please take a moment to have a look. I don't want to put it on hold for too long, otherwise have to do rebase later.

PavelVPV replied:
Sorry, a bit busy at the moment. Anyway, I'd like @alxelax and @omkar3141 to look at this PR before we merge it, but they will be back in January.

PavelVPV replied:
I didn't get your point. If the node has a proxy client connected to it and GATT Proxy feature is enabled, the non-filtered out messages will be relayed. When such node has prioritization enabled, this code will try to release the space in the pool, but will fail because of the additional reference in `proxy_msg.c`. So what will happen then is that such node will relay messages to the proxy client, but not over adv bearer.

LingaoM replied:
please see my current implementation, the proxy will use separate buffer.

LingaoM replied:
BTW： the proxy and friend relay message will use adv main pool instead of relay pool.

LingaoM commented on subsys/bluetooth/mesh/adv.h:
```diff
@@ -41,22 +36,41 @@ struct bt_mesh_adv {
+struct bt_mesh_buf {
+	sys_snode_t node;
+
+	struct {
```
No need

LingaoM commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:
```diff
@@ -659,6 +676,107 @@ static void test_tx_random_order(void)
+}
+
+static struct bt_mesh_send_cb first_relay_cb = {
+	.start = first_relay_send_start_cb,
```
yeah, Not used.

LingaoM commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -41,88 +41,157 @@ const uint8_t bt_mesh_adv_type[BT_MESH_ADV_TYPES] = {
+				    xmit, prio, K_NO_WAIT);
+#endif /* CONFIG_BT_MESH_RELAY */
+
+	return adv_create_from_pool(&adv_buf_pool,
```
This strange usage has a little history, but I found this:https://github.com/zephyrproject-rtos/zephyr/pull/29580

PavelVPV replied:
I need to look at this more detailly. I doubt if a node can relay a message if relay feature is disabled.

LingaoM replied:
Let's move to:https://github.com/zephyrproject-rtos/zephyr/pull/52072#discussion_r1043022353

LingaoM commented on tests/bluetooth/bsim_bt/bsim_test_mesh/tests_scripts/advertiser/tx_relay_order.sh:
```diff
@@ -0,0 +1,8 @@
+
+source $(dirname "${BASH_SOURCE[0]}")/../../_mesh_test.sh
+
+conf=prj_prio_relay_conf
```
Done.

PavelVPV commented:
> > It would be also ideal if the commit with switching to k_mem_slab would be added before the prioritization change.
> 
> I am thinking about the suitability of this matter, but it is also a thorny rebase for me. I think it makes no difference to the final result. Therefore, let's keep the status quo, but we can still do better next time, depending on more full and detailed preliminary consideration.

I think it shouldn't be difficult since eventual changes for prioritization are small. You could squash 2 commits, then pull out prioritization change, commit, and then add  prioritization change. Commit with test could just be moved to the end.

LingaoM committed:
Bluetooth: Mesh: Add relay table for relay logic

Reconstruct relay logic, Since Mesh Profile defines many types
of Relays, including Proxy and Friend, the previous Zephyr
implementations are confused, which is very annoying, so refactoring.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -41,88 +41,157 @@ const uint8_t bt_mesh_adv_type[BT_MESH_ADV_TYPES] = {
-		    adv_buf_destroy);
-
-static struct bt_mesh_adv adv_relay_pool[CONFIG_BT_MESH_RELAY_BUF_COUNT];
+K_MEM_SLAB_DEFINE(relay_buf_pool,
```
Done.

LingaoM commented on subsys/bluetooth/mesh/adv.h:
```diff
@@ -41,22 +36,41 @@ struct bt_mesh_adv {
+	uint8_t      prio;
+};
+
+struct bt_mesh_buf {
```
Yes, the current PR is enough, put it in another PR.

LingaoM replied:
@PavelVPV Thanks, finally have time to done now. It looks much better now

PavelVPV commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -648,101 +650,197 @@ static bool net_decrypt(struct bt_mesh_net_rx *rx, struct net_buf_simple *in,
+		return;
+	}
+
+	buf = bt_mesh_adv_main_create(BT_MESH_ADV_DATA,
```
I'm not sure I like this. Previously we didn't use an extra buffer. Now we don't only use an extra buffer, but also the local pool for relay messages. This will also occupy the whole local pool in high density traffic environment. And I think my comment https://github.com/zephyrproject-rtos/zephyr/pull/52072#discussion_r1052129083 is still relevant. This creates a situation where messages relayed to proxy, but not to adv bearer.

Maybe we could tag proxy packets somehow and then skip them?

LingaoM replied:
Well, I think most high density situations is adv to adv, not proxy, also for relay definitions, We confused these concepts before, I solved it by the way this time

LingaoM replied:
Also according https://github.com/zephyrproject-rtos/zephyr/pull/52072#discussion_r1044054258

The message delivered from adv bearer, if proxy support and enable, should relay to gatt bearer, this 'relay' not 'BT_MESH_RELAY', also shouldn't occupy relay buffer.

LingaoM replied:
BTW: the bt_mesh_relay function In the previous design, this function has always existed, even if there is no relay support.

LingaoM replied:
By the way, I remember that in the implementation of zephyr, the Bluetooth GATT connection also has a buffer, that is, it does not support intensive packet sending.

PavelVPV replied:
> Well, I think most high density situations is adv to adv, not proxy(because it's need keep a gatt connection), also for relay definitions, We confused these concepts before, I solved it by the way this time

If you have node A with gatt proxy connection and node B that sends or relays enormous amount of traffic, node A will relay those messages (assuming they bypassed all checks and filters). We can't disregard this case if prioritization feature is enabled on a node that supports GATT Proxy.

> The message delivered from adv bearer, if proxy support and enable, should relay to gatt bearer, this 'relay' not 'BT_MESH_RELAY', also shouldn't occupy relay buffer.

My main point is previously we were more efficient in terms of memory usage as we were using the same buffer and were just incrementing the reference. Now, as I mentioned above, we will use both pools. The spec doesn't care about buffers.

> the bt_mesh_relay function In the previous design, this function has always existed, even if there is no relay support.

I didn't get this point.

> By the way, I remember that in the implementation of zephyr, the Bluetooth GATT connection also has a buffer, that is, it does not support intensive packet sending.

I guess it doesn't matter as long as a user can configure the mesh with prioritization feature and GATT Proxy.

----

One more thing that came up to me regarding the previous code. In case when adv_ext.c relayed message before proxy, dereferencing the buffer will brake the code once the message is relayed by proxy. Since we can't cancel bt_gatt_notify_cb call, we can't touch that message, otherwise in proxy we will dereference either a new buf or buf with ref == 0.

Tbh at the moment I don't see a good way around this except either tagging messages relayed over gatt and skipping them (thought I expect it won't work if all messages are relayed over gatt), or making the prioritization and gatt proxy features mutually exclusive.

LingaoM replied:
> If you have node A with gatt proxy connection and node B that sends or relays enormous amount of traffic, node A will relay those messages (assuming they bypassed all checks and filters). We can't disregard this case if prioritization feature is enabled on a node that supports GATT Proxy.

Also have another solution, we could add new inner functions that send proxy relay messages without buf.

LingaoM replied:
> I didn't get this point.

I'm mean in the current implementation (main branch), even we do not support relay and proxy and friend, the bt_mesh_relay function also be compiled in final binary firmware.

PavelVPV replied:
> Also have another solution, we could add new inner functions that send proxy relay messages without buf.

Right, buf itself is not really needed for proxy, but we still need to notify upper layer about finishing transmission.

LingaoM replied:
One another way is send adv firstly, and send proxy relay message in adv:send_start.

LingaoM replied:
But will try add more args.

LingaoM replied:
Maybe can reuse buf.b.size or buf.b.__bufs

LingaoM replied:
@PavelVPV Ping

PavelVPV replied:
> One another way is send adv firstly, and send proxy relay message in adv:send_start.

I didn't get how this will help. We need to keep adv buf in pool until we finished transmission over GATT. If the adv buf that we want to remove should be relayed over gatt, and we scheduled its transmission, we can't remove it from the pool.

> Maybe can reuse buf.b.size or buf.b.__bufs

Could you elaborate this?

LingaoM replied:
I still stick to the existing solution(base on relay mapping table), because the Mesh Profile defines a set of relay concepts, including proxy and friend, and these concepts were confused in the previous code implementation, resulting in the current problem. The existing separation design is for the future The maintenance is better, even if it takes up one more buffer.

mia-ko commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -686,6 +686,19 @@ config BT_MESH_RELAY_BUF_COUNT
+config BT_MESH_RELAY_PRIORITY
+	bool "Support for prioritized relay messages"
+	help
+	  The prioritized relay messages will be send before other lower priority
+	  relay message to ensure that higher priority messages can be sent in time.
+	  When considering the message latency, also consider the values of
+	  BT_MESH_RELAY_RETRANSMIT_COUNT and BT_MESH_RELAY_RETRANSMIT_INTERVAL.
+	  A higher number of BT_MESH_RELAY_ADV_SETS allows the increase in the number
+	  of buffers while maintaining the latency.
+
+	  Note: To enable this option, the application layer needs to register a callback
+	  function to determine the priority of each relay message. If it is not registered,
+	  the default priority is the same as the lowest(zero).
```
```suggestion
	  Prioritized relay messages will be sent before other lower priority
	  relay messages. This ensures that the higher priority messages are sent in time.
	  When considering the message latency, also consider the values of
	  BT_MESH_RELAY_RETRANSMIT_COUNT and BT_MESH_RELAY_RETRANSMIT_INTERVAL.
	  A higher number of BT_MESH_RELAY_ADV_SETS allows the increase in the number
	  of buffers while maintaining the latency.

	  Note: To enable this option, the application layer needs to register a callback
	  function to determine the priority of each relay message. If it is not registered,
	  the default priority is lowest (zero).
```

LingaoM committed:
Bluetooth: Mesh: Rename struct buf to adv

`struct bt_mesh_adv` is more like about adv metadata and could
ideally be renamed to `bt_mesh_adv_meta`.

`struct bt_mesh_buf` could be renamed to `bt_mesh_adv` as it keeps
adv plus metadata

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
@PavelVPV @alxelax @omkar3141 Please take a look, thanks.

PavelVPV commented:
> @PavelVPV @alxelax @omkar3141 Please take a look, thanks.

Apologize for the delay. I'm a bit busy right now with some other stuff. I remember about this PR. I'll try to look at it next week. Thanks!

PavelVPV reviewed (commented):
One thing that we missed and found recently. Using multiple adv sets it is not guaranteed that the messages will be sent in the order they pushed to the adv sets. So even if we change order in the relay buffer, the messages can (and most likely will) be sent to the air in a wrong order.

PavelVPV commented:
> One thing that we missed and found recently. Using multiple adv sets it is not guaranteed that the messages will be sent in the order they pushed to the adv sets. So even if we change order in the relay buffer, the messages can (and most likely will) be sent to the air in a wrong order.

This reordering may not be a problem on ZLL when num_events = 1, but it happens with SDC in some mesh configurations.

LingaoM commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -34,112 +34,180 @@ LOG_MODULE_REGISTER(bt_mesh_adv);
+	int err;
 
-	net_buf_destroy(buf);
+	if (atomic_test_bit(bt_mesh.flags, BT_MESH_SUSPENDED)) {
```
Not really, it's actually be here already. Not changed this time.

LingaoM commented:
> > One thing that we missed and found recently. Using multiple adv sets it is not guaranteed that the messages will be sent in the order they pushed to the adv sets. So even if we change order in the relay buffer, the messages can (and most likely will) be sent to the air in a wrong order.
> 
> 
> 
> This reordering may not be a problem on ZLL when num_events = 1, but it happens with SDC in some mesh configurations.

Sure, the multiple advertising sets may schedule advertising events at link layers and not control by mesh layer, so it will happens, but I'm think not important.

LingaoM commented on tests/bluetooth/bsim_bt/bsim_test_mesh/overlay_prio_relay.conf:
```diff
@@ -0,0 +1,53 @@
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
+
+# Bluetooth configuration
+CONFIG_BT=y
```
Done. thanks/

PavelVPV reviewed (changes requested):
This change was discussed again within our team. As it was pointed earlier (https://github.com/zephyrproject-rtos/zephyr/pull/52072#issuecomment-1311915077) we still tend to think that we don't want to have the priorisation code in the upstream as is as it is not a Bluetooth mesh specification feature. But (as it was also pointed earlier) we are agree to have a hook in the advertiser bearer to pass the relay buffer to a user application and let the user reorder messages as they want. This will also allow to make the reordering more flexible (not just by priorities). And it seems to me this suggestion is not far from your current implementation.

The proxy issue (https://github.com/zephyrproject-rtos/zephyr/pull/52072#discussion_r1101458528) is still a problem and needs to be fixed. We should not use the main adv buffer for relaying messages over proxy.

LingaoM commented:
In our recent testing, the ability to prioritize relay significantly improved the control experience for devices that do not support extended advertising, although this PR closed.

Link another PR(separate local queue & relay queue for legacy adv) :https://github.com/zephyrproject-rtos/zephyr/pull/65043
